docs/source/coding_guidelines.md

```diff
@@ -0,0 +1,213 @@
+* Notebooks are also an easy way of leaking sensitive data: They can for example contain images derived from private datasets.
+
+## Review / Pull Requests
+
+### Scope of Pull Requests
+
+Pull Requests (PRs) should ideally implement a single change. If in doubt, err on the side of making the PR too small, rather than too big.
+
+* Small PRs help reduce the load on the reviewer.
+* Avoid adding unrelated changes to an existing PR.
+* PRs should be modular: we can iterate on PRs, and any positive delta is a contribution.
+
+Please follow the guidance on
+[Github flow](https://docs.github.com/en/get-started/quickstart/github-flow)
+
+Try gauging the value of your contribution for yourself by asking the following questions:
+
+* Will this change bring the team closer to achieving their project goals?
+* Will someone else understand my code?
+* Will they be able to use my code?
+* Will they be able to extend or build on top of my code?
+
+### Pull Request Process
+
+* When creating a Pull Request (PR), do add a summary of your contribution in the PR description.
+* The template PR description also contains a checklist for the PR author.
+* For collecting early feedback on your work, please use a Draft Pull Request. These PRs are marked with a grey icon in the Github UI, and send a clear signal that the code there is not yet ready for review. When submitting a draft PR, all the checks will be run.
+* Once your work is ready for review, click the "Ready for Review" button on the Github PR page, and assign reviewers for your PR.
+
+### Pull Request Contents
+
+* Include a brief description.
+* Link to an issue, if relevant.
+* Write unit tests for the code - see above for details.
+* Add appropriate documentation for any new code that you introduce.
+
+### Pull Request Titles
+
+To enable good auto-generated changelogs, we prefix all PR titles with a category string, like "BUG: Out of bounds error when using small images".
+Those category prefixes must be in upper case, followed by a colon (`:`). Valid categories are
+
+* `ENH` for enhancements, new capabilities
+* `BUG` for bugfixes
+* `STYLE` for stylistic changes (for example, refactoring) that do not impact the functionality
+* `DOC` for changes to documentation only
+* `DEL` for removing something from the codebase
+
+## Notes on Branching
+
+We should treat the main branch as a collection of code that is
+
+* Of high-quality
+* Readable and suitable for re-use
+* Well documented
+
+Use a dedicated dev branch for development, debugging and analysis work. Once you are sure that
+the work in the dev branch adds enough value to reach our project objectives, then use a pull request
+to get your contribution into the main branch.
+
+## Tools for code cleanup
+
+### Vulture
+
+[Vulture](https://pypi.org/project/vulture) is a tool to find orphaned functions.
+Use it occasionally to identify orphan functions and make sure that the repository remains simple.
```

Comment:
Can this be enforced in `pre-commit` instead?

Reply:
Yes, it could. At this point, I don't see value in automating vulture. What are your thoughts?

Reply:
I think vulture is something that Javier started to run on InnerEye, and subsequently also added to the coding guidelines doc. InnerEye was never really designed to be a re-usable library, so any function that was not used was a red flag. Now the situation is somehow different. 
We could add a recommendation for vulture on the project folders, but I'm not sure if it makes sense on something like hi-ml.

Reply:
Ok I think that makes sense.